src/libraries/System.Private.CoreLib/src/System/Resources/ResourceReader.Core.cs

```diff
@@ -65,24 +67,47 @@ private object DeserializeObject(int typeIndex)
             return graph;
         }
 
+        // TODO: Remove this DynamicDependencyAttributes when https://github.com/mono/linker/issues/943 is fixed.
+        [DynamicDependency("Deserialize", "System.Runtime.Serialization.Formatters.Binary.BinaryFormatter", "System.Runtime.Serialization.Formatters")]
         private void InitializeBinaryFormatter()
         {
-            LazyInitializer.EnsureInitialized(ref s_binaryFormatterType, () =>
-                Type.GetType("System.Runtime.Serialization.Formatters.Binary.BinaryFormatter, System.Runtime.Serialization.Formatters, Version=0.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a",
-                throwOnError: true)!);
-
-            LazyInitializer.EnsureInitialized(ref s_deserializeMethod, () =>
+            if (s_binaryFormatterType == null || s_deserializeMethod == null)
             {
-                MethodInfo binaryFormatterDeserialize = s_binaryFormatterType!.GetMethod("Deserialize", new Type[] { typeof(Stream) })!;
+                // Important: keep this as a local that is populated by a Type.GetType call so that the dataflow
+                // analysis in IL Linker can reason about the subsequent GetMethod call.
+                // Accessing the type from the s_binaryFormatterType field would make IL Linker lose track of the value
+                // (the value could be overwritten from another thread, so IL Linker can't make assumptions about it).
+                Type binaryFormatterType =
+                    Type.GetType("System.Runtime.Serialization.Formatters.Binary.BinaryFormatter, System.Runtime.Serialization.Formatters, Version=0.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a",
+                    throwOnError: true)!;
+
+                if (s_binaryFormatterType == null)
```

Comment:
Either delete the null check; or change this to `Interlocked.CompareExchange` to make sure the first one wins in thread-safe way.

Reply:
`Type.GetType` with the same string always returns the same instance, right? So deleting the null check should be fine.

Reply:
What is the thread safety concern? I got rid of the LazyInitializer because it looked like unnecessary ceremony for something that is inherently thread safe, but I would like to understand what it was trying to solve.

Reply:
Here was past discussion around this: https://github.com/dotnet/coreclr/pull/20907#discussion_r232264663

Reply:
> Type.GetType with the same string always returns the same instance, right?

Unless you manage to load multiple copies of the framework. We generally do not handle such situation well, so ignoring its existence should be ok.

> What is the thread safety concern?

My concern was that the null check as it is written right now provides false sense of thread-safety. It is either unnecessary or a thread safety bug.

I agree that LazyInitializer is way too expensive for that it does in 99% cases.